function_template.go

```diff
@@ -0,0 +1,78 @@
+package v8go
+
+// #include <stdlib.h>
+// #include "v8go.h"
+import "C"
+import (
+	"errors"
+	"runtime"
+	"unsafe"
+)
+
+// FunctionCallback is a callback that is executed in Go when a function is executed in JS.
+type FunctionCallback func(info *FunctionCallbackInfo) *Value
+
+// FunctionCallbackInfo is the argument that is passed to a FunctionCallback.
+type FunctionCallbackInfo struct {
+	ctx  *Context
+	args []*Value
+}
+
+// Context is the current context that the callback is being executed in.
+func (i *FunctionCallbackInfo) Context() *Context {
+	return i.ctx
+}
+
+// Args returns a slice of the value arguments that are passed to the JS function.
+func (i *FunctionCallbackInfo) Args() []*Value {
+	return i.args
+}
+
+// FunctionTemplate is used to create functions at runtime.
+// There can only be one function created from a FunctionTemplate in a context.
+// The lifetime of the created function is equal to the lifetime of the context.
+type FunctionTemplate struct {
+	*template
+}
+
+// NewFunctionTemplate creates a FunctionTemplate for a given callback.
+func NewFunctionTemplate(iso *Isolate, callback FunctionCallback) (*FunctionTemplate, error) {
```

Comment:
how does `vm.TerminateExecution` interact with callback of `v8go.NewFunctionTemplate`?

would it be possible to include `context.Context` as first argument of callback?

Reply:
Once the VM is terminated, you will no longer be able to fire the callback. These callbacks are connected to the isolate, so once the Isolate is GC'd so will be the callback functions (unless attached to some other struct etc)

You can get the current context via `info.Context()`

Reply:
`context.Context` will get confusing with `v8go.Context` (as I just did)
The V8 API does not have such a concept as the Go Context; if we introduce the Go Context, this would have to exist throughout the API, which would likely be a breaking change.

Reply:
I had a look at you code (https://github.com/choonkeat/try-v8go/blob/main/main.go)
I see what you are trying to do... you want to be able to check if the Go Context is `Done` within the callback?
Maybe we could attach a context in a way that is non-breaking ie. something like `info.GoContext()` ??

Reply:
@choonkeat Would you be able to submit this Issue as a feature request? It's maybe something we could look to supporting.

Reply:
> context.Context will get confusing with v8go.Context (as I just did)

ya :-P

> Would you be able to submit this Issue as a feature request?

it was more of a comment, since idiomatic Go callback function usually comes with `context.Context`. _But_ if it doesn't fit v8, I don't mind passing my own `context.Context` via closure

---

i'm still trying to wrap my head around how things "should" work. e.g. implementing `fetch` i'd need to return a Promise i think and thus will be tracking https://github.com/rogchap/v8go/issues/61 closely

Reply:
Yes, to implement true `fetch` you need to return the Promise; A poor man's version is here: https://github.com/rogchap/v8go/blob/master/function_template_test.go#L52

Reply:
FYI. this is better supported now that we have #76